### ISSUE TITLE
Dragging clips (Update Somehow Prevented It)

### ISSUE THREAD

**Issue by userofbugs** [OpenShot/openshot-qt#3290]
Hi, i'm a 'new' editor using openshot. This isn't my first time using the program but is my first time reporting a possible bug so let me explain...

I discovered what seems like a possible bug regarding clips being used. I'm not sure if it's part of the update to prevent clips from being moved around from its track or a randomly encountered bug but the newly released update has prevented me from 'drag & move' (which i like to call it) the clips that i wanted. I was able to drag and move my clip(s) around whether they were sliced or not freely before i had to update it.

So now... i'm stuck with clip(s) that i want removed from my tracks (which they aren't) but am unable to do so. The clips i wanted to edit in or out is a struggle to cope with now considering of the bug prevention.

Due to this report and bug, i'm delayed on finishing my editing and it pains me further to not being able to edit out what i don't want in the video.

In conclusion, the update somehow prevented me from freely moving my clips around (when they are on track) or even add them from under 'Project Files'. My temporary solution was to right-click on the files and click 'Add to timeline' which then adds the file's starting point on the right side of my playhead. 

So if anyone knows how to fix this, please let me know if you know how. Thank you

**Comment by ferdnyc** [comment]
@userofbugs 

Hmm, that's frustrating. Helpful information would be:

1. Your OS, OpenShot version installed (presumably 2.5.1, but release or Daily Build?), details like that.
2. A copy of your [`openshot-qt.log` file](https://github.com/OpenShot/openshot-qt/blob/develop/CONTRIBUTING.md) which may contain error information that could lead us to tracking down the problem. (Don't worry about following the collection procedure there, that document is just to help you locate the file. Whatever information we need would already be present in the existing log.)

I suspect the drag-and-drop handler has crashed in your running OpenShot session, which can happen on occasion for various reasons. (Like I said, hopefully there's info in the log that can tell us more.)

If that's the case, though, you _should_ be able to fix this by saving your current project, then quitting and restarting OpenShot. (You might want to save to a different file, rather than overwriting any previous save, just in case you need to recover the older project.) Typically when drag-and-drop goes haywire, it only lasts as long as the current session.

**Comment by userofbugs** [comment]
**OS**
Windows 10 Pro

**Current Version**
OpenShot-v2.5.1-x86_64.exe (Updated from 2.5.0)

**32bit Ver**
_OpenShot-v2.5.1-dev2-1583899607-4058dde3-232f3f79-x86.exe_ (the latest one i've downloaded, still not sure if i should use 32bit or 64bit)

**Explanation**
I currently have 2 bit versions (64 & 32) downloaded (just now downloaded 32bit cause i was shown to download the latest one) and the 64bit version is installed. What happens if 2 bit versions are both installed?

When i first deleted the first 2 files which was _openshot-qt.log_ & _libopenshot.log_  file, after that, i was trying to export my video but i realized a few seconds later... it crashed (multiple times with multiple attempts). Not so sure if deleting those 2 files were anything important after checking its content

Now that i uninstalled the 64bit ver and checked the 32bit wasn't installed as well to have no OpenShot installed to create no further confusion, i reinstalled the 64bit ver and now i want to know for sure if i again delete _openshot-qt-log & libopenshot.log_? won't be a problem... Also, how would i know if the name of _openshot-qt.log_ has a copy? Does it have the same name with '1 or 2' etc at the end of it?

Every time i re-open OpenShot (the updated 64bit), i opened up the project that i was working on, the clips that i said earlier wasn't resolved after closing the program.

Also, would it be a bad idea if i was to use the previous version (OpenShot-v.2.5.0) i was on before the update became an issue? I also want to know how to get to Preferences to enable 'debug mode'.

**Comment by SuslikV** [comment]
Please, share your project file (compress `.osp` into .zip format first), I think it is damaged. There is chance that your project can be repaired.

**Comment by ferdnyc** [comment]
> I currently have 2 bit versions (64 & 32) downloaded (just now downloaded 32bit cause i was shown to download the latest one) and the 64bit version is installed. What happens if 2 bit versions are both installed?

That shouldn't have any terribly adverse effects, other than possibly a little confusion about which version is being used to open project files, things like that. But they'll exist separately, one in "Project Files {x86)", the other in the 64-bit "Project Files". I know I've run with my Win7 VM set up like that on occasion.

> When i first deleted the first 2 files which was _openshot-qt.log_ & _libopenshot.log_ file, after that, i was trying to export my video but i realized a few seconds later... it crashed (multiple times with multiple attempts). Not so sure if deleting those 2 files were anything important after checking its content

There's definitely nothing OpenShot requires in the log files, it never even reads them — only writes to them. Deleting them should have absolutely no impact on OpenShot. (Unless, _possibly_, you delete them while OpenShot is running. I can't say I've tried that, so theoretically it could cause some issue. But that's the only thing I can think of that even _might_ be a problem.)

> Now that i uninstalled the 64bit ver and checked the 32bit wasn't installed as well to have no OpenShot installed to create no further confusion, i reinstalled the 64bit ver and now i want to know for sure if i again delete _openshot-qt-log & libopenshot.log_? won't be a problem... 

I don't _know of_ any reason it would cause problems, though as I say it's best to do only after quitting OpenShot completely.

> Also, how would i know if the name of _openshot-qt.log_ has a copy? Does it have the same name with '1 or 2' etc at the end of it?

Yes, exactly, you'd see `.openshot_qt.log.1`, `2`, etc. in the same directory.

> Every time i re-open OpenShot (the updated 64bit), i opened up the project that i was working on, the clips that i said earlier wasn't resolved after closing the program.

Drag-and-drop still wasn't working? Hmmm, that sounds like it could be a more serious issue with OpenShot's installation, then. It may be safest to completely delete the _entire_ `C:\Users\username\.openshot_qt\` directory. You'll lose your settings, but there aren't a whole lot of settings in OpenShot anyway, and it could be that something in that directory is causing problems. (You can always rename it to `.openshot_qt.old` or whatever, if you don't want to delete it right away. The point is to get OpenShot to create a whole new settings directory for you.)
 
> Also, would it be a bad idea if i was to use the previous version (OpenShot-v.2.5.0) i was on before the update became an issue? 

With 2.5.0, specifically? I would say yes. OpenShot 2.5.0 had something of a serious bug in how it managed file paths in the project file data, 2.5.1 was rushed out just to address that bug. (Which is why I'm thinking perhaps some leftovers of 2.5.0 in your settings directory are causing issues for you.) I would recommend downgrading to 2.4.4, if anything, rather than 2.5.0.

>I also want to know how to get to Preferences to enable 'debug mode'.

There's actually a Debug tab in the preferences, which has a checkmark to enable verbose logging. **_However_...** I would really recommend against it, in your case. The only effect of enabling debug mode in OpenShot is to turn on extensive logging (very, VERY extensive logging) from _libopenshot_, the library that handles the video processing for the application. It has absolutely **zero** effect on the user interface, except to make it operate more slowly (because of all the logging). So for things like drag-and-drop issues, turning on debugging only makes it harder to diagnose the issue, by burying the problem in tons of unrelated logging.

**Comment by userofbugs** [comment]
> 
> 
> Please, share your project file (compress `.osp` into .zip format first), I think it is damaged. There is chance that your project can be repaired.

I'm unsure if this is how you do it
I clicked on the .osp file that contained my work, then i right-clicked it to and scrolled for compress zip folder.

[ddalla.zip](https://github.com/OpenShot/openshot-qt/files/4318480/ddalla.zip)

**Comment by ferdnyc** [comment]
@userofbugs Yup, that worked perfectly thanks!

**Comment by ferdnyc** [comment]
@SuslikV Hrm. The very end of the file:
```json
  "redo": []
 },
 "version": {
  "openshot-qt": "2.5.1",
  "libopenshot": "0.2.5"
 },
 "4": null,
 "6": null,
 "5": null
}
```

(@userofbugs Basically, those "4", "5", "6" lines aren't supposed to be there... but it's an issue that keeps cropping up in users' project files.)

**Comment by userofbugs** [comment]
> 
> 
> > I currently have 2 bit versions (64 & 32) downloaded (just now downloaded 32bit cause i was shown to download the latest one) and the 64bit version is installed. What happens if 2 bit versions are both installed?
> 
> That shouldn't have any terribly adverse effects, other than possibly a little confusion about which version is being used to open project files, things like that. But they'll exist separately, one in "Project Files {x86)", the other in the 64-bit "Project Files". I know I've run with my Win7 VM set up like that on occasion.
> 
Ok... 2 different bit versions go into 2 separate files
> > When i first deleted the first 2 files which was _openshot-qt.log_ & _libopenshot.log_ file, after that, i was trying to export my video but i realized a few seconds later... it crashed (multiple times with multiple attempts). Not so sure if deleting those 2 files were anything important after checking its content
> 
> There's definitely nothing OpenShot requires in the log files, it never even reads them — only writes to them. Deleting them should have absolutely no impact on OpenShot. (Unless, _possibly_, you delete them while OpenShot is running. I can't say I've tried that, so theoretically it could cause some issue. But that's the only thing I can think of that even _might_ be a problem.)
> 
So those files will have no impact on OpenShot unless i have deleted those files with the program still open? 
> > Now that i uninstalled the 64bit ver and checked the 32bit wasn't installed as well to have no OpenShot installed to create no further confusion, i reinstalled the 64bit ver and now i want to know for sure if i again delete _openshot-qt-log & libopenshot.log_? won't be a problem...
> 
> I don't _know of_ any reason it would cause problems, though as I say it's best to do only after quitting OpenShot completely.
> 
I uninstalled and reinstalled of 2.5.0, 2.5.1 & 2.5.1 (32Bit) separately. Reopened each version and the effect is still the same with 'drag & drop' but even more painful, the export process crashes instantly which became strange cause it never happened before. It was a huge hassle to go through with but fortunately, the installing and uninstalling was fast enough.
> > Also, how would i know if the name of _openshot-qt.log_ has a copy? Does it have the same name with '1 or 2' etc at the end of it?
> 
> Yes, exactly, you'd see `.openshot_qt.log.1`, `2`, etc. in the same directory.
> 
Ah ok, are there any files within the directory i need to delete cause i don't really know which files should be or shouldn't be deleted apart from following the guide that lead me to downloading the 32bit version
> > Every time i re-open OpenShot (the updated 64bit), i opened up the project that i was working on, the clips that i said earlier wasn't resolved after closing the program.
> 
> Drag-and-drop still wasn't working? Hmmm, that sounds like it could be a more serious issue with OpenShot's installation, then. It may be safest to completely delete the _entire_ `C:\Users\username\.openshot_qt\` directory. You'll lose your settings, but there aren't a whole lot of settings in OpenShot anyway, and it could be that something in that directory is causing problems. (You can always rename it to `.openshot_qt.old` or whatever, if you don't want to delete it right away. The point is to get OpenShot to create a whole new settings directory for you.)
> 
I don't know about any 'settings' that i could've changed but i was using the user interface in a 'default' manner. I did change the way it looked when working within the user interface but nothing else more than that. 
Like for example
• properties on the right
• effects on the left
> > Also, would it be a bad idea if i was to use the previous version (OpenShot-v.2.5.0) i was on before the update became an issue?
> 
> With 2.5.0, specifically? I would say yes. OpenShot 2.5.0 had something of a serious bug in how it managed file paths in the project file data, 2.5.1 was rushed out just to address that bug. (Which is why I'm thinking perhaps some leftovers of 2.5.0 in your settings directory are causing issues for you.) I would recommend downgrading to 2.4.4, if anything, rather than 2.5.0.
> 
So the 2 latest versions (2.5.1 & 2.5.1 (32bit) are just rushed out to address the bugs that occured with 2.5.0?
> > I also want to know how to get to Preferences to enable 'debug mode'.
> 
> There's actually a Debug tab in the preferences, which has a checkmark to enable verbose logging. **_However_...** I would really recommend against it, in your case. The only effect of enabling debug mode in OpenShot is to turn on extensive logging (very, VERY extensive logging) from _libopenshot_, the library that handles the video processing for the application. It has absolutely **zero** effect on the user interface, except to make it operate more slowly (because of all the logging). So for things like drag-and-drop issues, turning on debugging only makes it harder to diagnose the issue, by burying the problem in tons of unrelated logging.

So debugging mode doesn't make the situation any better.

**Comment by userofbugs** [comment]
> 
> 
> @SuslikV Hrm. The very end of the file:
> 
> ```json
>   "redo": []
>  },
>  "version": {
>   "openshot-qt": "2.5.1",
>   "libopenshot": "0.2.5"
>  },
>  "4": null,
>  "6": null,
>  "5": null
> }
> ```
> 
> (@userofbugs Basically, those "4", "5", "6" lines aren't supposed to be there... but it's an issue that keeps cropping up in users' project files.)

Thank you for finding out. I did realize that there originally wasn't supposed to have more than 3 copies so it was half an unexpected surprise to find out that there's more popping out of nowhere but does having more than 3 copies mean anything crucial?

**Comment by SuslikV** [comment]
[ddalla - repaired.zip](https://github.com/OpenShot/openshot-qt/files/4318940/ddalla.-.repaired.zip)

about 16KB to download, CRC of the repaired .zip archive, MD5:55CD101F1295731899262B4FA19BF3E0

Try to unpack and place to the folder where your original project was. Should work well.

Technical details later.

**Comment by SuslikV** [comment]
@userofbugs Please, follow General Rules:

1. Make the backups of the project file time-to-time when you are working on something huge.
2. Do not update software until the project is done.

**Comment by SuslikV** [comment]
@ferdnyc I think this mysterious numbers as keys of dictionary can be set only in:

https://github.com/OpenShot/openshot-qt/blob/90e7bc4c9275f0b84340c6b4beda6e49540fee59/src/classes/project_data.py#L192

The project itself is not so huge, valid root keys:
```
{
"id":
"fps":
"display_ratio":
"pixel_ratio":
"width":
"height":
"sample_rate":
"channels":
"channel_layout":
"settings":
"clips":
"effects":
"export_path":
"import_path":
"files":
"duration":
"scale":
"tick_pixels":
"playhead_position":
"profile":
"layers":
"markers":
"progress":
"history":
"version":
}
```

**Comment by userofbugs** [comment]
> 
> 
> [ddalla - repaired.zip](https://github.com/OpenShot/openshot-qt/files/4318940/ddalla.-.repaired.zip)
> 
> about 16KB to download, CRC of the repaired .zip archive, MD5:55CD101F1295731899262B4FA19BF3E0
> 
> Try to unpack and place to the folder where your original project was. Should work well.
> 
> Technical details later.
> 
> **Edit:**
> Some `"markers"` becomes `null`, lines 24465-24467:
> Click to expand
> 
> the workaround was known long time ago: [#2430 (comment)](https://github.com/OpenShot/openshot-qt/issues/2430#issuecomment-446753224)
> 
> and as @ferdnyc said, the end of the file has additional dictionary keys that isn't required too. I removed them too.
> 
> This last part of the file is similar to other issues were clips doesn't moves after project file "damaged" with nulls (or with wrong values of the dictionary keys) - [#3258 (comment)](https://github.com/OpenShot/openshot-qt/issues/3258#issuecomment-592199925)
> 
> I personally didn't take any action to see why it happens in OpenShot.
> 
> I see, that you @ferdnyc looked at similar errors before: [#1051 (comment)](https://github.com/OpenShot/openshot-qt/issues/1051#issuecomment-488826066) and solution wasn't found yet (but some others were fixed).

Thank you for the repair. While you were repairing it, i tried opening other projects within OpenShot and the timeline of clips from those projects were unaffected. They were able to move without trouble. 

I also checked the locks to make sure they weren't enabled (so i wouldn't have made an issue about this) on the left side of its track _'Track 3'_ and to see if they weren't bugged from the last project i was doing. Turns out, it was that project alone `'ddalla.osp'` that was only affected. I did move clips from previously done projects and they worked like it wasn't bugged including the lock wasn't enabled to keep its clips immobilized.  

Now, i've unpacked and opened the project `ddalla-repaired.osp` in OpenShot, the 'drag & drop' issue has now seem to work better. Now after an attempt to export, i'm encountering an instant crash where it doesn't even reach `0.1%`. It seems to be a bigger issue that the exporting process is now `'unresponsive'`.

**Comment by SuslikV** [comment]
> ...i'm encountering an instant crash where it doesn't even reach `0.1%`

@userofbugs your PC just run out of memory. Build simple scene that suits both to your PC and to your demands.

**Comment by userofbugs** [comment]
> @userofbugs your PC just run out of memory. Build simple scene that suits both to your PC and to your demands.

My scenes aren't big per se but it does contain some `titles` above another track `track 4` & some video effects on the same track `track 3` or bottom on another track `track 2` of the main track `track 3`. 

I also did a 2 quick export test of a `few second short` video & same video that wasn't edited (not from the same project) and it can export any newly made video after that. So basically saying, OpenShot from what i can see can now export any new videos which means i can re-make the whole thing again

But from the sound of it, i probably need a new PC to do editing videos.

**Comment by ferdnyc** [comment]
> @ferdnyc I think this mysterious numbers as keys of dictionary can be set only in:
> 
> https://github.com/OpenShot/openshot-qt/blob/90e7bc4c9275f0b84340c6b4beda6e49540fee59/src/classes/project_data.py#L192

Yeah, looking at it I'm sure you're right. That code makes my brain sad, though. I have no idea _what_ it's doing TBH.

**Comment by ferdnyc** [comment]
> I also did a 2 quick export test of a `few second short` video & same video that wasn't edited (not from the same project) and it can export any newly made video after that. So basically saying, OpenShot from what i can see can now export any new videos which means i can re-make the whole thing again

Yeah, sorry about that. :disappointed: Unfortunately, sometimes a project file is just corrupted, and the best thing to do (if it's an option) is start over. The accumulated "distortion" (for lack of a better word) can be difficult to reverse.

Since you mentioned effects, though...
> My scenes aren't big per se but it does contain some `titles` above another track `track 4` & some video effects on the same track `track 3` or bottom on another track `track 2` of the main track `track 3`.

Before going _too_ far down the rebuild path, if you're using effects in your final output I'd recommend testing those early. There's new effects code in 2.5.1 that's _intended_ to make OpenShot render videos much faster, but it's certainly **possible** I messed something up there that's causing your crashes.

Another thing to watch out for: If there are any gaps in your project timeline where there are _no_ active clips at all — even for just a frame or two — some export codecs/formats will balk at that and abort the export process. There's an open bug about that somewhere, and an open feature request to build some sort of gap-checking or -closing tool, which is something we sorely need. But for the moment, I'm afraid the best solution is just to zoom in real close and make sure there aren't any unintended spaces between the different elements of your project.

**Comment by userofbugs** [comment]
> Another thing to watch out for: If there are any gaps in your project timeline where there are _no_ active clips at all — even for just a frame or two — some export codecs/formats will balk at that and abort the export process. There's an open bug about that somewhere, and an open feature request to build some sort of gap-checking or -closing tool, which is something we sorely need. But for the moment, I'm afraid the best solution is just to zoom in real close and make sure there aren't any unintended spaces between the different elements of your project.

If that's the case, how would i close the unnecessary gaps between the last clip of my `pre-export` video and the rest of its timeline (that has nothing else) including all the other tracks that have no active clips or effects whatsoever (for the tracks that i haven't used)?

**Comment by ferdnyc** [comment]
> including all the other tracks that have no active clips or effects whatsoever (for the tracks that i haven't used)?

Oh, it's not a per-track thing, only the _entire_ timeline. Any of the tracks can be empty at any point along the timeline, as long as (at least) _one_ of the tracks is not empty.

All tracks can be empty at the end of the timeline (they always will be, in fact), so that's not considered a gap. The timeline is sized dynamically, and its end point is defined as the last frame that contains an active clip — the space visible after that point is just "room to expand". (That dynamic sizing is why having a gap in the middle causes the export to abort prematurely — it _looks_ like the timeline has ended.)

So, when you export your pre-export video, it will be exported only up to the last active frame on the timeline. Clips by definition can't have gaps (a gap is defined as the absence of a clip), so when you import any video — including a pre-export — and place it on the timeline as  a clip, the clip length is exactly the length of the video content.

**Comment by ferdnyc** [comment]
(Also, only horizontal gaps matter — clips can be on the same track or different tracks, and they can be overlapping or adjacent (in terms of linear time).)

IOW:
![TimelineGaps](https://user-images.githubusercontent.com/538020/76853640-12199800-6824-11ea-854b-491cbf37eb63.png)

**Comment by userofbugs** [comment]
> 
> IOW:
> ![TimelineGaps](https://user-images.githubusercontent.com/538020/76853640-12199800-6824-11ea-854b-491cbf37eb63.png)

For the basics, i already knew i shouldn't have an active clip hanging around at the end of the timeline not being connected to the `second` last clip. Until 1 week ago is the only issue i had since i downloaded my first version of OpenShot (Version 2.4.4).

**Comment by SuslikV** [comment]
>... If there are any gaps in your project timeline where there are _no_ active clips at all — even for just a frame or two — some export codecs/formats will balk at that and abort the export process.

Never seen this before. @ferdnyc where is the mentioned issue or more info about?

**Comment by ferdnyc** [comment]
@SuslikV The export issue came up in #1984, looks like — it may only affect exports _to_ an image sequence, though, from reading it over again.

The open requests for a gap-management tool in general are #2808 / #3057.

**Comment by ferdnyc** [comment]
@userofbugs Apologies, I wasn't trying to imply... well, anything, really. A gap _anywhere_ on the timeline can be an issue, even between the *first* two clips (I just happened to illustrate it towards the end). It's actually pretty easy to accidentally create one of those during editing, as you're moving clips around and changing their length. (Especially if you work with the timeline view zoomed out, due to #2808.)

**Comment by userofbugs** [comment]
> @userofbugs Apologies, I wasn't trying to imply... well, anything, really. A gap _anywhere_ on the timeline can be an issue, even between the _first_ two clips (I just happened to illustrate it towards the end). It's actually pretty easy to accidentally create one of those during editing, as you're moving clips around and changing their length. (Especially if you work with the timeline view zoomed out, due to #2808.)

@ferdnyc No it's okay, i wasn't trying to sound offended and but all this conversation really helped me out with the actual issue i had. I will keep `A gap_anymore_on the timeline can be an issue` a priority when editing. Thanks very much!

@SuslikV Also thank you for repairing my file before.

I will report any other issues/bugs or maybe improvements if necessary when the time comes!
Appreciate the help :)

**Comment by ferdnyc** [comment]
Please do, absolutely!

The history on this is _really_ long, so in the interest of saving other project members from having to read it all, I'm going to close this even though there's still the unresolved project-corruption issue. (Which I'm confident @SuslikV has pinpointed the source of.)

Instead, I'll peel that comment off into a new issue, focused on that bug in the code.

Thanks for reporting!
